Mitsubishi AC decode

denxhun opened the pull request:
Mitsubishi AC decode implemented. Tested with unit test and with real remote controller.

denxhun committed:
Initial version of Mitsubishi AC decode

denxhun committed:
Remove comments disables full test cycle

denxhun committed:
Work in progres: Mitsubishi AC debugging.

denxhun committed:
Repeat handling enhanced for Mitsubishi AC

denxhun committed:
Final correction before PR

denxhun committed:
Merge branch 'master' into Mitsubishi-AC-decode

denxhun committed:
Dump demo output reverted.

denxhun committed:
Removed unnecessary include

crankyoldgit commented on examples/IRrecvDumpV2/IRrecvDumpV2.ino:
```diff
@@ -131,6 +132,16 @@ void dumpACInfo(decode_results *results) {
   }
 #endif  // DECODE_KELVINATOR
+#if DECODE_MITSUBISHI_AC
+  Serial.println("Trying Mitchu AC:");
```
Please remove the Serial.println()'s so it is consistent with the other decoders.

denxhun replied:
Sorry, left here from debug rounds.

crankyoldgit commented on src/ir_Mitsubishi.cpp:
```diff
@@ -292,6 +340,128 @@ void IRsend::sendMitsubishiAC(unsigned char data[], uint16_t nbytes,
+    }
+    if (!headerFound) {
+      DPRINTLN("Header mark not found.");
+      failure = true;
```
Why not just `return false;` instead?
Same for almost every following `failure = false;` too.
I'd suggest failing hard early, as it saves processing time on a wasted decode.

denxhun replied:
I use the `failure` because the repeat could help to interpret the messages - if we are not in strict mode. If I return here then I can't benefit from repeat.

crankyoldgit replied:
Fair enough. Though I think relying on the repeat to decode successfully is a bit redundant. We should get it the first time.

crankyoldgit commented on src/ir_Mitsubishi.cpp:
```diff
@@ -292,6 +340,128 @@ void IRsend::sendMitsubishiAC(unsigned char data[], uint16_t nbytes,
+      }
+    }  // strict repeat check
+  } while (failure && rep <= MITSUBISHI_AC_MIN_REPEAT);
+  results->decode_type = MITSUBISHI_AC;
```
Need to also record the number of bits as per every other decoder.

denxhun replied:
Valid.

denxhun committed:
Revert "Removed unnecessary include"

This reverts commit a44045b829c3a9cac8ec9b7213a51ce26b84a3ba.

denxhun committed:
Corrections applied after PR comments.

crankyoldgit commented:
Also, given the "2.5-dev" in the branch name, I've moved this PR to point to the 2.5-dev branch. That will mean you probably need to rebase you're commits.

crankyoldgit commented on src/ir_Mitsubishi.cpp:
```diff
@@ -292,6 +343,129 @@ void IRsend::sendMitsubishiAC(unsigned char data[], uint16_t nbytes,
+// Ref: https://www.analysir.com/blog/2015/01/06/reverse-engineering-mitsubishi-ac-infrared-protocol/
+bool IRrecv::decodeMitsubishiAC(decode_results *results, uint16_t nbits,
+                              bool strict) {
+  if (results->rawlen < ((MITSUBISHI_AC_STATE_LENGTH * 8 * 2) + 2)) {
```
MITSUBISHI_AC_STATE_LENGTH * 8 is MITSUBISHI_AC_BITS
Use that instead

denxhun replied:
You are right, I can modify to `MITSUBISHI_AC_STATE_LENGTH * 2` (because mars and spaces).

crankyoldgit commented on src/ir_Mitsubishi.cpp:
```diff
@@ -292,6 +343,129 @@ void IRsend::sendMitsubishiAC(unsigned char data[], uint16_t nbytes,
+//  Somtime happens that junk signals arrives before the real message
+    bool headerFound = false;
+    while (!headerFound && offset <
+        (results->rawlen - (MITSUBISHI_AC_STATE_LENGTH * 8 * 2 + 2))) {
```
MITSUBISHI_AC_BITS instead again.

denxhun replied:
Right.

crankyoldgit commented on src/ir_Mitsubishi.cpp:
```diff
@@ -292,6 +343,129 @@ void IRsend::sendMitsubishiAC(unsigned char data[], uint16_t nbytes,
+    if (rep != MITSUBISHI_AC_MIN_REPEAT && failure) {
+      bool repeatMarkFound = false;
+      while (!repeatMarkFound && offset <
+          (results->rawlen  - (MITSUBISHI_AC_STATE_LENGTH * 8 * 2 + 4))) {
```
Ditto

denxhun replied:
Yep.

crankyoldgit commented on src/ir_Mitsubishi.cpp:
```diff
@@ -350,14 +524,25 @@ uint8_t* IRMitsubishiAC::getRaw() {
+
 // Calculate the checksum for the current internal state of the remote.
 void IRMitsubishiAC::checksum() {
+  remote_state[17] = calculateChecksum(remote_state);
```
Nice cleanup of the checksum stuff!

denxhun replied:
Thanks.

crankyoldgit commented on src/ir_Mitsubishi.cpp:
```diff
@@ -85,6 +88,54 @@ void IRsend::sendMitsubishi(uint64_t data, uint16_t nbits, uint16_t repeat) {
+//
+// Status: DEAT / Appears to be working.
+//
+bool IRrecv::matchByte(uint8_t *resultByte, volatile uint16_t *data_ptr,
```
We should probably make `matchData()` handle LSB data rather than add another routine.

denxhun replied:
You are right. I didn't want to modify such a widely used function for my needs. This is a more light weight solution but if you prefer I can move this functionality to there.

crankyoldgit replied:
I think it's best if we can. I use it elsewhere and then do `reverseBits()` when I want it in LSB order. Doing it as part of the match is a better idea.

Don't worry about breaking it. That's what the unit tests are for. They'll catch most things if you break it. ;)

crankyoldgit replied:
I've upgraded `matchData()` to handle LSBF bit ordering. So this function can go and convert to use the standard `matchData()` function.

crankyoldgit commented on src/ir_Mitsubishi.cpp:
```diff
@@ -430,11 +615,21 @@ uint8_t IRMitsubishiAC::getMode() {
-    case MITSUBISHI_AC_DRY: break;
-    case MITSUBISHI_AC_HEAT: break;
-    default: mode = MITSUBISHI_AC_AUTO;
+    case MITSUBISHI_AC_AUTO:
```
Why do this? Why not change the constants to 0b0011xxxx instead?
Also, Why the extra 0b0011? Did we miss it in the initial decoding analysis?

denxhun replied:
The mode setting wasn't perfectly implemented before. The mode is stored in 2 places of the payload and represented differently. Byte 6 and byte 8. Reference: www.analysir.com/blog/wp-content/uploads/2014/12/Mitsubishi_AC_IR_Signal_Fields.jpg
I checked that my remote controller does exactly this.

crankyoldgit replied:
Ah and D'oh! I missed it was [8] in the switch statement, and [6] afterwards.

crankyoldgit commented on test/ir_Mitsubishi_test.cpp:
```diff
@@ -689,11 +689,283 @@ TEST(TestMitsubishiACClass, MessageConstuction) {
+  398,  438, 400,  448, 390,  446, 392,  444, 394,  444, 394,  442, 396,
+  440, 398,  438, 400,  438, 400,  448, 392,  446, 392,  444, 394,  442, 396,
+  442, 396,  440, 398,  438, 1240,  444, 1264,  422, 390,  446, 392,  446,
+// Repeat mark (1111) wrong:
```
Maybe we should increase the tolerance value so 1111 matches.

denxhun replied:
This is intentionally wrong. Here I test what happen when my decoder failed to find the repeat mark&space.

crankyoldgit reviewed (commented):
I should have said this first:

Thank you for doing this, and it's excellent stuff. Doing a AC decoder/encoder is hard. I'm impressed.

crankyoldgit committed:
Merge branch 'v2.5-dev' into v2.5-dev-mitsubishi-ac-decode

crankyoldgit committed:
Changes to make it use 2.5-dev'isms.

crankyoldgit committed:
Fix a lint error.

crankyoldgit commented on src/ir_Mitsubishi.cpp:
```diff
@@ -84,6 +87,54 @@ void IRsend::sendMitsubishi(uint64_t data, uint16_t nbits, uint16_t repeat) {
+//   zerospace:   desired length os the space for 0.
+//   MSBFirst:    most significant bit first? (default false).
+//
+// Status: DEAT / Appears to be working.
```
BETA

crankyoldgit commented on src/ir_Mitsubishi.cpp:
```diff
@@ -3,6 +3,9 @@
 
```
Please add yourself here, in ir_Mitsubishi_test.cpp, and in https://github.com/markszabo/IRremoteESP8266/blob/master/.github/Contributors.md too.

You've earned it!

crankyoldgit replied:
I've added you to the ir_Mitsubishi* files.

crankyoldgit committed:
Extend matchData() to handle excess & diff bit ordering.
Should be 100% backward compatible with all previous invocation instances.
Added Unit test coverage for different bit ordering.

crankyoldgit committed:
Fix a linter/style issue.

crankyoldgit committed:
Replace matchByte() with matchData()
Remove matchByte()

crankyoldgit committed:
Add denxhun to copyright info & code style improvements

denxhun commented:
Hi David,

Many thanks for your effort in merging my far-from-perfect PR.

crankyoldgit commented:
@denxhun No worries. If I waited for perfect PRs, I'd never merge any of my crappy code. ;-)

crankyoldgit commented:
@denxhun FYI v2.5.0 has been released which includes your code. ;-)

denxhun commented:
Thanks a lot. I'm so proud to be part of it.

crankyoldgit commented:
So you should be. Remember: Contributions & assistance are _always_ welcome at any time.
